### ISSUE TITLE
Maria DB version issue in dialects/mysql/base.py

### ISSUE THREAD

**Issue by jmbski** [sqlalchemy/sqlalchemy#9885]
### Describe the bug

I ran into an error this morning where my python server started throwing fits about some value called @@transaction_isolation not existing. After some research, I discovered it's due to something about the mysql version for Maria DB 9.6.0 (and presumably others) returning a value that doesn't match up properly. Discussion thread on that [here](https://code.djangoproject.com/ticket/28804#comment:6). 

I did also read [this discussion thread](https://github.com/sqlalchemy/sqlalchemy/issues/5583) that touched on the subject.

It was working fine previously, I think it may have been caused by my web hosting provider updating my DB recently (I also got a notification on MySQL workbench). Not sure, but I only noticed it this morning.

I decided to take a crack at fixing it locally and made a change that worked (for me). I changed [this bit of code here](https://github.com/sqlalchemy/sqlalchemy/blob/d9d0ffd96c632750be9adcb03a207d75aecaa80f/lib/sqlalchemy/dialects/mysql/base.py#L2495) from:
``` 
        if self._is_mysql and self.server_version_info >= (5, 7, 20):
            cursor.execute("SELECT @@transaction_isolation")
        else:
            cursor.execute("SELECT @@tx_isolation")

to:

```
            if self._is_mysql:
                try:
                    cursor.execute("SELECT @@transaction_isolation")
                except Exception as inst:
                    print(inst)
                    print("@@transaction_isolation failed, trying @@tx_isolation")
                    try:
                        cursor.execute("SELECT @@tx_isolation")
                    except Exception as inst:
                        print("@@tx_isolation also failed, logging error")
                        print(inst)

I thought about doing a pull request etc.. but I'm not as familiar with y'all's procedures for that and code formatting preferences etc... so I thought it'd be easier to just raise the issue, show the solution I came up with, and let you guys handle it as you will. 

I figured that a try/except block would be more robust, and that since the original code was going to execute one of the two options either way, try/excepting both wouldn't be a problem. But I'm not as familiar with the infrastructure so take my opinion with a grain of salt.

Not sure if repro code is exactly what you're looking for, let me know if I need to provide anything else.

### SQLAlchemy Version in Use

2.01 and 2.0.15

### DBAPI (i.e. the database driver)

pymysql ( I think )

### Database Vendor and Major Version

MySQL 8 (I believe), MariaDB 9.6.0

### Python Version

3.11

### Operating system

Ubuntu 22.04.2 LTS

### To Reproduce

```python
from sqlalchemy import create_engine

def get_mysql_engine(config_file):
    # Initialize a dictionary to store the connection settings
    connection_settings = {}

    # Read the config file
    config_file_data = open(config_file).readlines()

    # Extract the connection settings from the config file
    for line in config_file_data:
        if "DB_HOST" in line:
            connection_settings["host"] = line.split("=")[1].strip()
        elif "DB_PORT" in line:
            connection_settings["port"] = line.split("=")[1].strip()
        elif "DB_USER" in line:
            connection_settings["user"] = line.split("=")[1].strip()
        elif "DB_PASSWORD" in line:
            connection_settings["password"] = line.split("=")[1].strip()
        elif "DB_NAME" in line:
            connection_settings["dbname"] = line.split("=")[1].strip()

    db_host = connection_settings["host"]
    db_name = connection_settings["dbname"]
    db_user = connection_settings["user"]
    db_password = connection_settings["password"]
    db_port = connection_settings["port"]
    print(
        f"mysql+mysqlconnector://"
        + db_user
        + ":"
        + db_password
        + "@"
        + db_host
        + ":"
        + db_port
        + "/"
        + db_name
    )

    # I believe the initial entry point for the break is here
    engine = create_engine(
        f"mysql+pymysql://{db_user}:{db_password}@{db_host}/{db_name}"
    )

    # Return the connection settings
    return engine
#flask decorators etc..
def fake_rest_api_endpoint():
    
    engine = get_mysql_engine("properties/environment.properties")
    #normally the rest of my REST call follows, and works fine, but it breaks here and
    #nothing else gets executed without the fix I applied (or an equivalent)
```

### Error

```
Traceback (most recent call last):
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3293, in raw_connection
    return self.pool.connect()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 1268, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 168, in _do_get
    with util.safe_reraise():
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 166, in _do_get
    return self._create_connection()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
    return _ConnectionRecord(self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
    self.__connect()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 916, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1915, in go
    return once_fn(*arg, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/create.py", line 733, in first_connect
    dialect.initialize(c)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/dialects/mysql/base.py", line 2717, in initialize
    default.DefaultDialect.initialize(self, connection)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 527, in initialize
    self.default_isolation_level = self.get_default_isolation_level(
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 578, in get_default_isolation_level
    return self.get_isolation_level(dbapi_conn)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/dialects/mysql/base.py", line 2437, in get_isolation_level
    cursor.execute("SELECT @@transaction_isolation")
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/cursors.py", line 158, in execute
    result = self._query(query)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/cursors.py", line 325, in _query
    conn.query(q)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 549, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 779, in _read_query_result
    result.read()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 1157, in read
    first_packet = self.connection._read_packet()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 729, in _read_packet
    packet.raise_for_error()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1193, "Unknown system variable 'transaction_isolation'")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 2548, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.handle_exception(e)
  File "/home/joseph/.local/lib/python3.10/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 1822, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/joseph/.local/lib/python3.10/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 1820, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/joseph/.local/lib/python3.10/site-packages/flask/app.py", line 1796, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/joseph/.local/lib/python3.10/site-packages/flask_cors/decorator.py", line 128, in wrapped_function
    resp = make_response(f(*args, **kwargs))
  File "/home/joseph/coding_base/dnd-battle-sim/backend/src/main.py", line 68, in register_user
    Base.metadata.create_all(engine)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/sql/schema.py", line 5796, in create_all
    bind._run_ddl_visitor(
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3243, in _run_ddl_visitor
    with self.begin() as conn:
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3233, in begin
    with self.connect() as conn:
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3269, in connect
    return self._connection_cls(self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 147, in __init__
    Connection._handle_dbapi_exception_noconnection(
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2431, in _handle_dbapi_exception_noconnection
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3293, in raw_connection
    return self.pool.connect()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 1268, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 168, in _do_get
    with util.safe_reraise():
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 166, in _do_get
    return self._create_connection()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
    return _ConnectionRecord(self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
    self.__connect()
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 916, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1915, in go
    return once_fn(*arg, **kw)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/create.py", line 733, in first_connect
    dialect.initialize(c)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/dialects/mysql/base.py", line 2717, in initialize
    default.DefaultDialect.initialize(self, connection)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 527, in initialize
    self.default_isolation_level = self.get_default_isolation_level(
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 578, in get_default_isolation_level
    return self.get_isolation_level(dbapi_conn)
  File "/home/joseph/.local/lib/python3.10/site-packages/sqlalchemy/dialects/mysql/base.py", line 2437, in get_isolation_level
    cursor.execute("SELECT @@transaction_isolation")
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/cursors.py", line 158, in execute
    result = self._query(query)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/cursors.py", line 325, in _query
    conn.query(q)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 549, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 779, in _read_query_result
    result.read()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 1157, in read
    first_packet = self.connection._read_packet()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/connections.py", line 729, in _read_packet
    packet.raise_for_error()
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/home/joseph/.local/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1193, "Unknown system variable 'transaction_isolation'")
``` 

### Additional context
Added note: this is my first time ever submitting an issue ticket  like this on a public github repo, so if I've flubbed any of the expected parts, please be kind lol

I think I've about covered everything. If you know of a reason a try/except wouldn't work or have a smarter way to fix it I'd be fine with that. For now I'm going to continue using my local variation, but when I move this up to my proper web hosting server I'd prefer not to have to do any weird shenanigans to use an alternate version for a couple line change. 

Or if there's something else I can do to work around it without adjusting the source. I'm not picky, I just want my web app to work is all. Any help/advice would be appreciated! Thanks in advance.

**Comment by CaselIT** [comment]
Hi,

I'm not familiar with that mariadb version. https://mariadb.com/kb/en/mariadb-releases/
If it's version 10.6, that's strange since it's one of the versions tested in the ci.

Can you please re-check?

**Comment by jmbski** [comment]
> Hi,
> 
> I'm not familiar with that mariadb version. https://mariadb.com/kb/en/mariadb-releases/ If it's version 10.6, that's strange since it's one of the versions tested in the ci.
> 
> Can you please re-check?

that's what it's saying it is. Screenshot from phpMyAdmin. Additionally I did a print of the self.server_version_info and it came back as (9,6,0). I could be misreading which one of them is 9.6.0, I'm not super familiar with the inner details of MySQL beyond general SQL/DB usage. In any case, that server version info comes back as (9,6,0) and attempts to use @@transaction_isolation instead of tx and then throws an error that doesn't get caught

![Screenshot_20230602-090332-313](https://github.com/sqlalchemy/sqlalchemy/assets/54730166/69907d72-21f4-484a-9b3c-dfd2da3f3a00)

**Comment by zzzeek** [comment]
hi -

this is some kind of version detection thing, so we need to know exactly what version you are on.

please log into your database and show us the complete output of "SELECT VERSION()":

```sql
$ mysql -u scott test -p
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 6
Server version: 10.5.19-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [test]> SELECT VERSION();
+-----------------+
| VERSION()       |
+-----------------+
| 10.5.19-MariaDB |
+-----------------+
1 row in set (0.000 sec)

```

**Comment by jmbski** [comment]
> hi -
> 
> this is some kind of version detection thing, so we need to know exactly what version you are on.
> 
> please log into your database and show us the complete output of "SELECT VERSION()":
> 
> ```sql
> $ mysql -u scott test -p
> Enter password: 
> Welcome to the MariaDB monitor.  Commands end with ; or \g.
> Your MariaDB connection id is 6
> Server version: 10.5.19-MariaDB MariaDB Server
> 
> Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.
> 
> Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
> 
> MariaDB [test]> SELECT VERSION();
> +-----------------+
> | VERSION()       |
> +-----------------+
> | 10.5.19-MariaDB |
> +-----------------+
> 1 row in set (0.000 sec)
> ```

Alright got a chance to get back to my computer. SELECT VERSION(); responded with 9.6.0 as well. I ran it via MySQL Workbench and phpMyAdmin. 

As a note, my python server is hosted locally at the moment, the DB is hosted on a server I have with Host Gator. As such, I have limited access to the actual server itself to run commands, if that makes a difference.

**Comment by zzzeek** [comment]
> Alright got a chance to get back to my computer. SELECT VERSION(); responded with 9.6.0 as well. I ran it via MySQL Workbench and phpMyAdmin.

hi,

please paste the **complete output**, I need **exactly** what the output was, tokens like "MariaDB" being present or not are of particular importance, this is how we differentiate here and is likely where things are going wrong so I need the **complete version string copy-and-pasted exactly**, thanks

**Comment by jmbski** [comment]
> > Alright got a chance to get back to my computer. SELECT VERSION(); responded with 9.6.0 as well. I ran it via MySQL Workbench and phpMyAdmin.
> 
> hi,
> 
> please paste the **complete output**, I need **exactly** what the output was, tokens like "MariaDB" being present or not are of particular importance, this is how we differentiate here and is likely where things are going wrong so I need the **complete version string copy-and-pasted exactly**, thanks

that is the complete output as far as I can get it. I'm investigating at the moment if there's a way for me to SSH to the actual server, but I don't know if that's possible

**Comment by zzzeek** [comment]
is there a server administrator or someone that admins this DB?   it seems like you are on some kind of mysql variant (there are many).

**Comment by jmbski** [comment]
> is there a server administrator or someone that admins this DB? it seems like you are on some kind of mysql variant (there are many).

I'll check. I think their customer service can provide that for me. I'll update here in a bit once I find out more information

**Comment by zzzeek** [comment]
here let's make this easy, run this program, after substituting your local login values:

```py

import pymysql

# substitute your values here
db_user = "scott"
db_password="tiger"
db_host="localhost"
db_name="test"

conn = pymysql.connect(user=db_user, passwd=db_password, host=db_host, db=db_name)
cursor = conn.cursor()
cursor.execute("SELECT VERSION()")
print(cursor.fetchall())


```

**Comment by jmbski** [comment]
> here let's make this easy, run this program, after substituting your local login values:
> 
> ```python
> import pymysql
> 
> # substitute your values here
> db_user = "scott"
> db_password="tiger"
> db_host="localhost"
> db_name="test"
> 
> conn = pymysql.connect(user=db_user, passwd=db_password, host=db_host, db=db_name)
> cursor = conn.cursor()
> cursor.execute("SELECT VERSION()")
> print(cursor.fetchall())
> ```

sorry for the late response, the rest of my day ended up being busy. So I ran this and still got the same:
```
(('9.6.0',),) 

is the exact copy paste of the output. I'm also currently on the line with customer support to have it run locally on the server just in case. 

Additional note, for me personally, this won't be an issue much longer as I switched to a different hosting platform (Hostinger) that has a full control VPS for a fraction of what I was paying Host Gator. But I've still got the other one online for a while longer if you want to continue troubleshooting the versioning issue. As a fellow developer (though not as much with python and DB connection stuff like sqlalchemy), I'm assuming the worry is that this particular bug is just a symptom of a potentially bigger problem with getting version info and not just a one off edge case?

**Comment by CaselIT** [comment]
The issue here seems that this particular server is reporting an unknown versuon, since there is no version 9 of mysql or mariadb.

So features for this server are a bit of a mistery

**Comment by jmbski** [comment]
> The issue here seems that this particular server is reporting an unknown versuon, since there is no version 9 of mysql or mariadb.
> 
> So features for this server are a bit of a mistery

I was wondering if it might be something along those lines. I saw that mine said it was 9.6 and checked online for information about it and got nothing. Wasn't sure if I just wasn't looking hard enough or it was a mystery version lol

**Comment by zzzeek** [comment]
this issue seems to have died down, and we have no way to fix on this end since "9.1.6" is not a mariadb version string, so closing.
